domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -486,38 +471,80 @@ class TopicListController @Inject constructor(
     val topicIdJsonArray = jsonAssetRetriever
       .loadJsonFromAsset("topics.json")!!
       .getJSONArray("topic_id_list")
+
+    // The list of started or completed topic IDs.
+    val startedTopicIds = topicProgressList.map(TopicProgress::getTopicId)
+    // All topics that could potentially be recommended.
     val topicIdList =
       (0 until topicIdJsonArray.length()).map { topicIdJsonArray[it].toString() }
-
+    // The list of topic IDs that qualify for being recommended.
+    val unstartedTopicIdList = topicIdList.filterNot { startedTopicIds.contains(it) }
+
+    // A map of topic IDs to their dependencies.
+    val topicDependencyMap = topicIdList.associateWith {
+      retrieveTopicDependencies(it).toSet()
+    }.withDefault { setOf<String>() }
+    // The list of topic IDs that are considered "finished" from a recommendation perspective.
+    val fullyCompletedTopicIds = topicProgressList.filter {
+      topicHasAtLeastOnStoryCompleted(it)
+    }.map(TopicProgress::getTopicId)
+    // A set of topic IDs that can be considered topics that should not be recommended.
+    val impliedFinishedTopicIds = computeImpliedCompletedDependencies(
+      fullyCompletedTopicIds, topicDependencyMap
+    )
     // Suggest prerequisite topic user needs to learn after completing any of the topics.
-    for (i in 0 until topicIdList.size) {
-      val topicInProgressFound = topicProgressList.any { it.topicId == topicIdJsonArray[i] }
-      if (!topicInProgressFound) {
-        val dependentListOfTopics =
-          retrieveTopicDependencies(topicIdList[i])
-        for (j in dependentListOfTopics.indices) {
-          topicProgressList.mapIndexed { index, topicProgress ->
-            if (topicProgress.topicId == dependentListOfTopics[j]) {
-              val topic = topicController.retrieveTopic(topicProgressList[index].topicId)
-              if (checkIfAtLeastOneStoryIsCompleted(topicProgressList[index], topic)) {
-                val recommendedTopicAlreadyExist =
-                  recommendedStories.any { it.topicId == topicIdList[i] }
-                if (!recommendedTopicAlreadyExist) {
-                  val recommendedStoriesIdFromAssets =
-                    createRecommendedStoryFromAssets(topicIdList[i])
-                  if (recommendedStoriesIdFromAssets != null) {
-                    recommendedStories.add(recommendedStoriesIdFromAssets)
-                  }
-                }
-              }
-            }
-          }
+    for (topicId in unstartedTopicIdList) {
+      // All of the topic's prerequisites must be completed before it can be suggested.
+      val dependentTopicIds = topicDependencyMap[topicId]
```

Comment:
```suggestion
      val dependentTopicIds = topicDependencyMap[topicId] ?: listOf()
```

We should ensure non-nullness here since unknown dependencies can sort of sensibly be defaulted to no dependencies.

Reply:
Done.